### PR TITLE
Add persist options for temporal store

### PR DESCRIPTION
**Pull request by SugarF0x**
Expose internal temporal store state creator function to allow for middleware injections which include but are not limited to persist
-----
Todo:
- [X] Cleanup
- [X] Fix tests

### CODE REVIEW

**Comment by charkour** [comment]
Thanks for this! @SugarF0x

I'll help update the tests. Would you like me to make a PR in your fork, is it okay to push to this branch (you might have to click "allow edits from maintainers".

**Comment by SugarF0x** [comment]
@charkour Are, ive checked maintainers' edits so you can push straight to my fork's main branch

**Comment by charkour** [comment]
Sorry, there was a bug due to the lastest version of zustand. Would you be willing to update this PR?

**Comment by SugarF0x** [comment]
@charkour aye, fixed that up nicely, working just as fine 👌

**Comment by SugarF0x** [comment]
@charkour reconsidered approach, allowed use of practically any kind and any amount of temporal store middlewares while containing the logic within the temporal middleware itself 👌 
now just gotta figure out a proper type cast for it 🤔

**Comment by charkour** [comment]
Awesome! I'll take a look at it now. Thanks for the work on it!

**Comment by charkour** [comment]
@SugarF0x, do you like the ability to add different middleware to the temporal store than the main store?

Originally, I thought that we could add the temporal state to the main store. What do you think? Maybe both should be an option.

**Comment by SugarF0x** [comment]
@charkour we could add an option to toggle between the in-store and out-of-store options to base the temporal store at

Initially i also thought about adding temporal to the main store but then decided that this would be a bit of a breaking change

Plus, Im not too convinced that tampering with one's store directly is not too user friendly since that happens implicitly

Imagine adding an `undo` function to the parent store directly where user has already had an `undo` function declared effectively overriding it making one question why their function does not what they want it to

Change merge order and now we have a different issue - user declared `undo` function overrides temporal `undo` function making it impossible to call temporal undo at all

While that could be pointed out in the README as a caveat or all the temporal functionality could be shoved under a `_temporal` key at the store root level, but is it really that much better?

I dont think either option is inherently good or bad, Im just trying to play devils advocate here

Both ideas are good and the final choice is on you as the author

In the end, i only ever stuck with the current implementation of having temporal in a dedicated store purely out of the backwards-compatibility consideration

**Comment by charkour** [comment]
I agree with you and your line of thinking. Let's go forward with keeping the temporal state separate from the user state, but I'll also add a separate function that can be used to inject the temporal state into the user state to make middleware simpler.

**Comment by charkour** [comment]
@SugarF0x, would you be willing to update the tests for this PR? I can't seem to be able to edit this branch. 

I'm working on the function that allows you to inject directly into the user's state.

**Comment by SugarF0x** [comment]
weird 🤔 

image

coming back to the backwards compatibility, no prior tests break 👐 

image

**Comment by charkour** [comment]
Tests are looking good!

Thanks for leading this feature development! I think once this is merged, we can release v2 finally!

I'll check this over in a couple hours.

File: packages/zundo/src/temporal.ts
Comment (charkour) [inline packages/zundo/src/temporal.ts:22]: Thanks for inlining these. Was meaning to do that.

File: apps/web/pages/subscribeWithSelector.tsx
Comment (charkour) [inline apps/web/pages/subscribeWithSelector.tsx:3]: Does subscribeWithSelector still work with your updates?

File: packages/zundo/src/types.ts
Comment (charkour) [inline packages/zundo/src/types.ts:31]: What other names did you think of for storeWrap? I'm wondering if there is a better one we could use. 

Perhaps, temporalMiddleware or simply middleware? Maybe storeMiddleware or wrapTemporalStore?

I'm not sure if there are other examples is in the zustand ecosystem that use this pattern we could copy. 

What do you think?

**Review by charkour — COMMENTED** [review]
A couple quick comments

File: packages/zundo/src/index.ts
Comment (charkour) [inline packages/zundo/src/index.ts:71]: Is this export needed?

Comment (SugarF0x) [reply packages/zundo/src/index.ts:71]: ideally we should not be exposing that, its currently sitting here for testing purposes

File: packages/zundo/src/types.ts
Comment (SugarF0x) [reply packages/zundo/src/types.ts:31]: storeWrap was more of a WIP TBD name, either `middlewares` or `middlewareInjection` would be much more suitable

alternatively, this could be called `storeInitializerInjection` since the purpose if it is to add some custom functionality e.g. middlewares to a store initializator which is the zustand name for the function we pass to `create` method

File: apps/web/pages/subscribeWithSelector.tsx
Comment (SugarF0x) [reply apps/web/pages/subscribeWithSelector.tsx:3]: good question, i just assumed that since tests still do pass, everything should be fine
didn't look into it myself tho 🤔

Comment (SugarF0x) [reply apps/web/pages/subscribeWithSelector.tsx:3]: plust this was an empty page so i just plugged it to avoid compilation errors for the time being

Comment (charkour) [reply apps/web/pages/subscribeWithSelector.tsx:3]: If you'd be willing to restore this manual test, that would be great!

I couldn't think of a easy way to unit test this. Something to add in the future.

File: packages/zundo/src/types.ts
Comment (charkour) [reply packages/zundo/src/types.ts:31]: Okay, I prefer the name `wrapTemporal` or `wrapTemporalStore`. 

What do you think?

Once these small comments are addressed, we can merge this in, add docs, and release a new version! Thank you for all the great help.

Comment (SugarF0x) [reply packages/zundo/src/types.ts:31]: Happy to be working on this, amazing opportunity to both learn more about middlewares as well as to do some good for the people using this much like i do

`wrapTemporalStore` sounds good to me

**Review by charkour — APPROVED** [review]
This is looking good! Thanks for this. 

Once this is no longer a draft, we can merge it in!

**Comment by SugarF0x** [comment]
ye i just gotta figure out how to fix the typing system
its quite complicated with how zustand handles middleware state interface merging
all looks fine on paper but in practice the types go haywire 🤯

**Comment by charkour** [comment]
Sounds good! The types stretch my understanding as well. 

I'll look into the type errors as well.

**Comment by SugarF0x** [comment]
@charkour i fixed the typings
does it work? yes
is it pretty? debatable

**Comment by SugarF0x** [comment]
decided to check how other middleware libraries handle middleware injection

image

💀

**Comment by charkour** [comment]
Oh wow. Not much type safety there.

**Comment by charkour** [comment]
Also, the typings are looking good! Thanks again for working on it.

**Comment by charkour** [comment]
Would you like a review on this? Or is it still a draft?

**Comment by SugarF0x** [comment]
@charkour ye i think its review worthy at this point
i wanna give the tests another look this weekend but the implementation itself is pretty much ready i believe

File: packages/zundo/vitest.config.ts
Comment (charkour) [inline packages/zundo/vitest.config.ts:6]: This is a nice package. Helpful for testing. Thanks!

**Comment by charkour** [comment]
Thanks for the updates to the branch!

**Comment by charkour** [comment]
Yeah, sorry this is taking so long. Let's say one week. That way I'll prioritize reviewing this.

Thank you so much for your effort and patience.

**Comment by charkour** [comment]
I'll merge this into the `rc-2` branch to prep the Release Candidate for v2

**Comment by charkour** [comment]
Thanks for your patience on this.

**Pull request merged**
